runtime/tr.source/trj9/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -5381,6 +5381,257 @@ bool genInstanceOfOrCheckCastNullTest(TR::Node* node, TR::CodeGenerator* cg, TR:
+   /* Layout of the writable data snippet
+    * Case - 1 : Cast class is runtime variable
+    * [UpdateIndex][ObjClassSlot-0][CastClassSlot-0]...[ObjClassSlot-N][CastClassSlot-N]
+    * Case - 2 : Cast Class is inetrface / unresolved
+    * [UpdateIndex][ObjClassSlot-0]...[ObjClassSlot-N]
+    * If there is only one cache slot, we will not have header.
+    * Last bit of cached objectClass will set to 1 indicating false cast
+    */
+   int32_t snippetSizeInBytes = ((cacheCastClass ? 2 : 1) * maxOnsiteCacheSlots * addressSize) + (addressSize * (maxOnsiteCacheSlots != 1));
+   if (needToGenerateDynamicCacheTest)
+      {
+      TR::S390WritableDataSnippet *dynamicCacheSnippet = NULL;
+      /* We can only request the snippet size of power 2, following table summarizes bytes needed for corresponding number of cache slots
+       * Case 1 : Cast class is runtime variable
+       * Case 2 : Cast class is interface / unresolved
+       * Number Of Slots |  Bytes needed for Case 1 | Bytes needed for Case 2
+       *        1        |              16          |           8
+       *        2        |              64          |           32
+       *        3        |              64          |           32
+       *        4        |              128         |           64
+       *        5        |              128         |           64
+       *        6        |              128         |           64
+       */
+      int32_t requestedBytes = 1 << (int) (log2(snippetSizeInBytes-1)+1);
+      traceMsg(comp, "Requested Bytes = %d\n",requestedBytes);
+      uintptr_t *initialSnippet = new (cg->trHeapMemory()) uintptr_t[requestedBytes/sizeof(uintptr_t)];
+      // NOTE: For single slot cache, we initialize snippet with addressSize (4/8) assuming which can not be objectClass
+      // In all cases, we use first addressSize bytes to store offset of the circular buffer and rest of buffer will be initialized with 0xff.
+      initialSnippet[0] = addressSize;
+      if (requestedBytes > addressSize)
+         memset(initialSnippet+1,-1,((requestedBytes/sizeof(uintptr_t))-1)*sizeof(initialSnippet));
+      dynamicCacheSnippet = (TR::S390WritableDataSnippet*)cg->CreateConstant(node, initialSnippet, requestedBytes, true);
+      
+      int32_t currentIndex = maxOnsiteCacheSlots > 1 ? addressSize : 0;
+      dynamicCacheReg = srm->findOrCreateScratchRegister();
+      /* Inlined tests for instanceOf node is generated such that, fall through of all test is next test. So we do not need to generate a dynamic cache test label. 
+       * Only exception to this is Super Class test with dynamic cast class. As we don't know dynamic cast class at compile time, we need to check type of cast class at runtime. 
+       * If it is regular class, super class test should yield to result other wise it will branch to dynamic cache test if dynamic cache is enabled.
+       */
+      if (dynamicCacheTestLabel != NULL)
```

Comment:
Do we need all this complexity though? We could just always generate this label. It doesn't affect the generated binary at all and it may be simpler in terms of code maintenance.